docs/architecture/adr-105-refactor-mempool-senders.md

```diff
@@ -0,0 +1,183 @@
+	reqRes.SetCallback(mem.reqResCb(tx, txInfo, cb))
+```
+When we send a request for validating a transaction via `CheckTxAsync`, it
+returns a `ReqRes` object. To handle the response asynchronously, we set an
+ad-hoc callback `reqResCb` on `ReqRes`. This callback is different for each
+transaction `tx` because it's parameterized by `tx`, `txInfo` (which essentially
+contains the sender ID), and another callback function `cb` provided by the
+caller of `CheckTx` to be applied on the response.
+
+Secondly, the list of senders for each transaction is recorded directly in the
+mempool's `txs` data structure. However, the list of senders is an important
+component of the propagation protocol and it should be part of the reactor,
+while `txs` is part of the implementation of this specific version (v0) of the
+mempool.
+
+In this document, we propose a solution that involves moving the list of senders
+from the mempool implementation to the reactor. This change will simplify the
+code, establish a more clear separation of the propagation protocol and the data
+structure, and allow for future improvements to the mempool as a whole.
+
+## Detailed Design
+ 
+We propose the following changes to the mempool's reactor and the `CListMempool`
+implementation.
+
+- In the `Mempool` interface, change the signature of `CheckTx` from
+    ``` golang
+    CheckTx(tx types.Tx, cb func(*abci.ResponseCheckTx), txInfo TxInfo) error
+    ```
+    to
+    ``` golang
+    CheckTx(tx types.Tx) (abcicli.ReqRes, error)
+    ```
+  - The returning `ReqRes` object can be used to set and invoke a callback to
+    handle the response, if needed.
+  - The callback parameter `cb` is no longer needed. Currently, this is mainly
+    used by the RPC endpoints `broadcast_tx_sync` and `broadcast_tx_commit`, and
+    in tests for checking that the response is valid. However, the same
+    functionality can be obtained with `ReqRes` response.
+  - `txInfo` contains information about the sender, which is also no longer
+    needed, as justified by the next point.
+- The list of senders for each transaction is currently stored in `mempoolTx`,
+  the data structure for the entries of `txs`. Move the senders out of
+  `mempoolTx` to a new map `txSenders` of type `map[types.TxKey]map[uint16]bool`
+  in the mempool reactor. `txSenders` would map transaction keys to a set of
+  peer ids (of type `uint16`). Add also a `cmtsync.RWMutex` lock to handle
+  concurrent accesses to the map.
+  - This refactoring should not change the fact that the list of senders live as
+    long as the transactions are in the mempool. When a transaction is received
+    by the reactor (either via RPC or P2P), we call `CheckTx`. We know whether a
+    transaction is valid and was included in the mempool by reading the `ReqRes`
+    response. If this is the case, record the list of senders in `txSenders`.
+    When a transaction is removed from the mempool, notify the reactor to remove
```

Comment:
Should remove ~, with the channel described below.~ since the channel is one of the options.